plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -360,6 +361,196 @@ struct block_time_tracker {
+      // initial vote_track_mode from its default value of `vote_track_mode::disabled`
+      // to `vote_track_mode::inactive` so that a future call to `update_active_finalizers`
+      // has an opportunity to set it correctly going forward.
+      if (!producers.empty() && production_pause_vote_timeout.count() != 0) {
+         _vt.set_vote_timeout(production_pause_vote_timeout);
+         _vote_track_mode.store(vote_track_mode::inactive, std::memory_order_relaxed);
+      }
+   }
+
+   enum class vote_track_mode {
+      disabled,
+      inactive,
+      only_other_votes,
+      only_producer_votes,
+      all_votes,
+   };
+
+   vote_track_mode get_vote_track_mode() const {
+      return _vote_track_mode.load(std::memory_order_relaxed);
+   }
+
+   static bool is_active(vote_track_mode vtm) {
+      return vtm != vote_track_mode::disabled && vtm != vote_track_mode::inactive;
+   }
+
+   // Called on resume()
+   void force_unpause() {
+      if (!is_active(get_vote_track_mode()))
+         return;
+
+      _vt.force_unpause(); // safe to always call, but no need if not active
+   }
+
+   auto check_pause_status(fc::time_point now) const {
+      const auto vtm = get_vote_track_mode();
+
+      production_pause_vote_tracker::pause_check check = production_pause_vote_tracker::pause_check::both;
+
+      switch (vtm) {
+      case vote_track_mode::disabled:
+      case vote_track_mode::inactive:
+         return production_pause_vote_tracker::pause_status{}; // should_pause() will return false
+      case vote_track_mode::only_other_votes:
+         check = production_pause_vote_tracker::pause_check::other;
+         break;
+      case vote_track_mode::only_producer_votes:
+         check = production_pause_vote_tracker::pause_check::producer;
+         break;
+      case vote_track_mode::all_votes:
+         check = production_pause_vote_tracker::pause_check::both;
+         break;
+      }
+
+      return _vt.check_pause_status(now, check);
+   }
+
+   // called from multiple threads
+   void on_vote(uint32_t connection_id, vote_result_t status, const vote_message_ptr& msg,
+                const finalizer_authority_ptr& active_auth, const finalizer_authority_ptr& pending_auth)
+   {
+      if (!is_active(get_vote_track_mode()))
+         return;
+
+      switch( status ) {
+      case vote_result_t::success:
+      case vote_result_t::duplicate:
+         break;
+      case vote_result_t::unknown_public_key:
+      case vote_result_t::invalid_signature:
+      case vote_result_t::max_exceeded:
+      case vote_result_t::unknown_block:
+         return;
+      default:
+         assert(false); // should never happen
+      }
+
+      if (!active_auth && !pending_auth) {
+         fc_elog(_log, "vote signal contains no valid authority ${m}", ("m", msg));
+         return;
+      }
+
+      const auto now = fc::time_point::now();
+      account_name auth_desc = active_auth ? to_account_name_safe(active_auth->description)
```

Comment:
Probably `auth_desc` should be called `finalizer` or `finalizer_name`.

Reply:
I changed it to `finalizer_auth_desc`. It is `finalizer_authority.description`.